GenTs.cs

```diff
@@ -0,0 +1,297 @@
+//
+// Copyright (C) 2011-2020  Piotr Fusik
+//
+// This file is part of CiTo, see https://github.com/pfusik/cito
+//
+// CiTo is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// CiTo is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with CiTo.  If not, see http://www.gnu.org/licenses/
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Foxoft.Ci
+{
+  /**
+   * TypeScript code generator
+   *
+   * At the moment, this only generates TypeScript declarations (.d.ts files) which don't contain method bodies.
+   * In the future we could implement full TS source code generation.
+   */
+  public class GenTs : GenJs
+  {
+
+    void Write(CiEnum enu)
+    {
+      // WARNING: TypeScript enums allow reverse lookup that the Js generator currently
+      // doesn't implement
+      WriteLine();
+      Write(enu.Documentation);
+      Write("export enum ");
+      Write(enu.Name);
+      OpenBlock();
+      int i = 0;
+      foreach (CiConst konst in enu)
+      {
+        Write(konst.Documentation);
+        WriteUppercaseWithUnderscores(konst.Name);
+        Write(" = ");
+        if (konst.Value != null)
+          konst.Value.Accept(this, CiPriority.Statement);
+        else
+          Write(i);
+        WriteLine(',');
+        i++;
+      }
+      CloseBlock();
+    }
+
+    protected override void WriteTypeAndName(CiNamedValue value)
+    {
+      WriteName(value);
+      Write(": ");
+      Write(value.Type);
+    }
+
+    void Write(CiType type)
+    {
+      switch (type)
+      {
+        case null:
+          Write("void");
+          break;
+        case CiNumericType _:
+          Write("number");
+          break;
+        case CiStringType _:
+          Write("string");
+          break;
+        case CiEnum enu:
+          Write(enu.Name);
+          break;
+        case CiListType list:
+          Write(list.ElementType);
+          Write("[]");
+          break;
+        case CiSortedDictionaryType dict:
+          Write("Record<");
+          Write(dict.KeyType);
+          Write(", ");
+          Write(dict.ValueType);
+          Write('>');
+          break;
+        case CiDictionaryType dict:
+          Write("Record<");
+          Write(dict.KeyType);
+          Write(", ");
+          Write(dict.ValueType);
+          Write('>');
+          break;
+        case CiArrayType array:
+          Write(array.ElementType);
+          Write("[]");
+          break;
+        default:
+          if (type == CiSystem.MatchClass)
```

Comment:
Currently only `Match()` (object storage, not a pointer) is supported, which is not legal in a method signature.

Reply:
I wrote this part looking at the same thing in `GenJava`...will it eventually be possible for functions to accept `Match#` parameters or return `Match#`?

Reply:
ad17d80fbb115baa8ed3d18b3f98526920b7ff38 makes read-only `Match` references legal.
`Match!` and `Match#` are disallowed, because they don't map well to C#, where the match gets [assigned](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.match?view=netcore-3.1#System_Text_RegularExpressions_Regex_Match_System_String_System_String_).